### ISSUE TITLE
[RepoHashAlgorithm] undefined reference error which is caused by incorrect hash caculation.

### ISSUE THREAD

**Issue by MaggieYingYi** [SNSystems/llvm-project-prepo#110]
Consider the following IR file.

$ cat test.ll
~~~LLVM
target triple = "x86_64-pc-linux-gnu-elf"

%"T1" = type { i8* }
%"T2" = type { i32 (...)**, i32, i8* }

@Var1 = internal constant { [1 x i8*] } { [1 x i8*] [i8* bitcast (i8* (%T1*, i8*)* @Func1 to i8*)] }, align 8

@Var2 = global %"T2" zeroinitializer, align 8

@llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 65535, void ()* @_GLOBAL__sub_I_test.cpp, i8* null }]

define internal i8* @Func1(%T1* %this, i8* %R) {
entry:
  %R.addr = alloca i8*, align 16
  %R2 = getelementptr inbounds %T1, %T1* %this, i32 0, i32 0
  store i8* %R, i8** %R2, align 16
  ret i8* %R
}

define internal void @__cxx_global_var_init() {
entry:
  call void @Func2(%"T2"* @Var2)
  ret void
}

define internal void @Func2(%"T2"* %this) {
entry:
  %0 = bitcast %"T2"* %this to i32 (...)***
  store i32 (...)** bitcast (i8** getelementptr inbounds ({ [1 x i8*] }, { [1 x i8*] }* @Var1, i32 0, inrange i32 0, i32 2) to i32 (...)**), i32 (...)*** %0, align 8
  ret void
}

define internal void @_GLOBAL__sub_I_test.cpp() {
entry:
  call void @__cxx_global_var_init()
  ret void
}
~~~

Compile this and dump the fragment corresponding to `Var2`.
~~~
$ rm clang.db
$ clang -c -O3 --target=x86_64-pc-linux-gnu-repo -x ir test.ll -o repo.o
$ pstore-dump --fragment=$(python fragments.py -d repo.o Var2) clang.db
---

- file      :
      path : clang.db
      size : 4194304
  fragments :
      - digest   : 242b89b38e6c24288e4f9535cdeb35c0
        fragment :
            - type     : data
              contents :
                  align   : 8
                  data    : !!binary |
                      AAAAAAAAAAA=
                  ifixups : [ ]
                  xfixups :
                      - name   : Var1
                        type   : 1
                        offset : 0
                        addend : 16
...
~~~
We can see the `Var1` is a xfixup of `Var2`.
Now, dump the compilation members in the repo.o file. There are 3 compilation members.
~~~
$ pstore-dump --compilation=$(repo-ticket-dump repo.o) clang.db
- file         :
      path : clang.db
      size : 4194304
  compilations :
      - digest      : 43f54b2433d6f1b5df82aff1b95b8586
        compilation :
            members :
                - digest     : 8ba0f0bae1396ddab5d086bfb66d7dde
                  fext       : { addr: 544, size: 55 }
                  name       : Func1
                  linkage    : internal
                  visibility : default
                - digest     : 5aaaba418d5cb1880ea77e7fea1c13ab
                  fext       : { addr: 416, size: 88 }
                  name       : Var1
                  linkage    : internal
                  visibility : default
                - digest     : 7508099180f3e80ae77fb5ce1c4b49fa
                  fext       : { addr: 640, size: 104 }
                  name       : Var2
                  linkage    : external
                  visibility : default
            path    : \\\\?\\E:\\MyWork\\Repo_Bugs\\NewIssue\\case4
            triple  : x86_64-pc-linux-gnu-repo
~~~
Now, I modified the `Func1` by changing the instruction alignment.
~~~
define internal i8* @Func1(%T1* %this, i8* %R) {
entry:
  %R.addr = alloca i8*, align 8
  %R2 = getelementptr inbounds %T1, %T1* %this, i32 0, i32 0
  store i8* %R, i8** %R2, align 8
  ret i8* %R
}
~~~
Compile the file again and dump the fragment corresponding to `Var2`.
~~~
$ clang -c -O3 --target=x86_64-pc-linux-gnu-repo -x ir test.ll -o repo1.o
$ pstore-dump --fragment=$(python fragments.py -d repo1.o Var2) clang.db
---
- file      :
      path : clang.db
      size : 4194304
  fragments :
      - digest   : 7508099180f3e80ae77fb5ce1c4b49fa
        fragment :
            - type     : data
              contents :
                  align   : 8
                  data    : !!binary |
                      AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
                  ifixups : [ ]
                  xfixups :
                      - name   : Var1
                        type   : 1
                        offset : 0
                        addend : 16
...
~~~
We can see the fragment keep the same. However, if we dump the corresponding compilation file, we can see the fragment `Var1` is not in the compilation file. This will results in the undefined reference in the linking stage.
~~~
$ pstore-dump --compilation=$(repo-ticket-dump repo1.o) clang.db
---
- file         :
      path : clang.db
      size : 4194304
  compilations :
      - digest      : b7bddac947d5b76e91aca61cee5ac91f
        compilation :
            members :
                - digest     : 7508099180f3e80ae77fb5ce1c4b49fa
                  fext       : { addr: 640, size: 104 }
                  name       : Var2
                  linkage    : external
                  visibility : default
            path    : \\\\?\\E:\\MyWork\\Repo_Bugs\\NewIssue\\case4
            triple  : x86_64-pc-linux-gnu-repo
...
~~~

**Comment by MaggieYingYi** [comment]
### Triage:

If you dump the digest of `Var2` in two object files, you could see they are the same.
~~~
$ python fragments.py -d repo.o Var2 clang.db
7508099180f3e80ae77fb5ce1c4b49fa
$ python fragments.py -d repo1.o Var2 clang.db
7508099180f3e80ae77fb5ce1c4b49fa
~~~
Debug the repo hash algorithm using the following commands:
~~~
$ rm clang.db
$ opt -S -debug-only prepo-digest -mtriple x86_64-pc-linux-gnu-repo test.ll -o repo.ll

GO Name: Func1
GOInfo:
        Initial Digest: b523143711b7893d634b24651fab812b
        Contributions: [ ]
        Dependencies: [ ]

GO Name:llvm.global_ctors
GOInfo:
        Initial Digest: f4b02192b126ab5ae338f773208f4806
        Contributions: [ ]
        Dependencies: [ _GLOBAL__sub_I_test.cpp]

GO Name: __cxx_global_var_init
GOInfo:
        Initial Digest: f95d50eb95e160638c8925ceba6c3a6c
        Contributions: [ ]
        Dependencies: [ Var2, Func2]

GO Name: Func2
GOInfo:
        Initial Digest: b637bc57e5559f717e6c66fececdeeb3
        Contributions: [ ]
        Dependencies: [ Func1, Var1]

GO Name: Var1
GOInfo:
        Initial Digest: 166f26f4eaa8544aae40f83e3bb0274e
        Contributions: [ ]
        Dependencies: [ Func1]

GO Name: Var2
GOInfo:
        Initial Digest: c75870dd5b16700e95ab61f6473c4cdc
        Contributions: [ __cxx_global_var_init]
        Dependencies: [ ]

GO Name: _GLOBAL__sub_I_test.cpp
GOInfo:
        Initial Digest: 2716b660189b757bfd39def23d6ac488
        Contributions: [ ]
        Dependencies: [ __cxx_global_var_init]
~~~
We could see the `Var2` only has a contribution `__cxx_global_var_init`. However, `__cxx_global_var_init` has the dependencies `Var2, Func2`. The function `Func2` has the dependencies `Func1, Var1`. Therefore, the hash change of `Func1` results in the changes of `__cxx_global_var_init`, which results in the hash change of `Var2`. Therefore, the digests of the `Var2` in two object files should be different.

The current hash algorithm hasn’t caught the change.

**Comment by MaggieYingYi** [comment]
### Reason:

The current hash algorithm considers `Var2` to be cached. Dump the debug information using the following command, we can see the `Var2` is cached.
~~~
$ opt -S -debug-only prepo-digest -mtriple x86_64-pc-linux-gnu-repo test.ll -o repo.ll
Computing hash for "Func1" (#0)
Recording result for "Func1"
Computing hash for "__cxx_global_var_init" (#0)
Computing hash for "Var2" (#1)
Recording result for "Var2"
Computing hash for "Func2" (#2)
Computing hash for "Func1" (#3)
Returning pre-computed hash for "Func1"
Computing hash for "Var1" (#3)
Computing hash for "Func1" (#4)
Returning pre-computed hash for "Func1"
Recording result for "Var1"
Recording result for "Func2"
Recording result for "__cxx_global_var_init"
Computing hash for "Func2" (#0)
Returning pre-computed hash for "Func2"
Computing hash for "_GLOBAL__sub_I_test.cpp" (#0)
Computing hash for "__cxx_global_var_init" (#1)
Returning pre-computed hash for "__cxx_global_var_init"
Recording result for "_GLOBAL__sub_I_test.cpp"
Computing hash for "Var1" (#0)
Returning pre-computed hash for "Var1"
Computing hash for "Var2" (#0)
Returning pre-computed hash for "Var2"
Computing hash for "llvm.global_ctors" (#0)
Computing hash for "_GLOBAL__sub_I_test.cpp" (#1)
Returning pre-computed hash for "_GLOBAL__sub_I_test.cpp"
Recording result for "llvm.global_ctors"
~~~

This is caused by the commit https://github.com/SNSystems/llvm-project-prepo/commit/7b7f92cd717b48aaba33916f7f46ad5356141ad0 . The function ` updateDigestUseContributions` only considers the GO’s contributions. It should consider the dependencies of the GO’s contributions as well.

**Comment by MaggieYingYi** [comment]
Fixed in commit 7e93bf0266.
